File: 프로그래머스/2/42584. 주식가격/주식가격.py
from collections import deque

def solution(prices):
    q = deque()
    
    answer = [0] * len(prices)
    
    # 값, index
    for i in range(len(prices)):
        while q and q[-1][0] > prices[i]:
            price, index = q.pop()
            answer[index] = i - index
        q.append([prices[i], i])

    
    while q:
        index = q.popleft()[1]
        answer[index] = len(prices) - 1 - index
        
    return answer

File: 프로그래머스/2/42584. 주식가격/주식가격.java
import java.util.*;

class Solution {
    public int[] solution(int[] prices) {
        int[] answer = new int[prices.length];

        Deque<int[]> stack = new LinkedList<>();
        
        for (int i = 0; i < prices.length; i++) {
            while(!stack.isEmpty() && stack.peekLast()[0] > prices[i]) {
                int[] xi = stack.pollLast();
                answer[xi[1]] = i - xi[1];
            }
            stack.addLast(new int[]{prices[i], i});
        }
        
        for (int[] xi : stack) {
            answer[xi[1]] = prices.length - xi[1] - 1;
        }
        
        return answer;
    }
}